src/meepgeom.cpp

```diff
@@ -1145,6 +1188,77 @@ void geom_epsilon::eff_chi1inv_matrix(meep::component c, symmetric_matrix *chi1i
+  double rad;            // (spherical) voxel radius
+  double uval;           // bilinearly-interpolated weight at voxel center
+  double ugrad_abs;      // magnitude of gradient of uval
+  double beta;           // thresholding bias
+  double eta;            // thresholding erosion/dilation
+  vector3 med1_eps_diag; // diagonal of epsilon tensor from medium 1
+  vector3 med2_eps_diag; // diagonal of epsilon tensor from medium 2
+};
+
+#ifdef CTL_HAS_COMPLEX_INTEGRATION
+static cnumber matgrid_ceps_func(int n, number *x, void *mgva_) {
+  matgrid_volavg *mgva = (matgrid_volavg *)mgva_;
+  // use a linear approximation for the material grid weights around the Yee grid point
+  double u_proj = tanh_projection(mgva->uval + mgva->ugrad_abs*x[0], mgva->beta, mgva->eta);
+  vector3 med1_eps_diag = mgva->med1_eps_diag;
+  vector3 med2_eps_diag = mgva->med2_eps_diag;
+  double eps1 = (med1_eps_diag.x + med1_eps_diag.y + med1_eps_diag.z)/3;
+  double eps2 = (med2_eps_diag.x + med2_eps_diag.y + med2_eps_diag.z)/3;
+  cnumber ret;
+  ret.re = (1-u_proj)*eps1 + u_proj*eps2;
+  ret.im = (1-u_proj)/eps1 + u_proj/eps2;
+  double w = 0;
+  if (mgva->dim == meep::D1)
+    w = 1/(2*mgva->rad);
+  else if (mgva->dim == meep::D2)
+    w = 2*sqrt(mgva->rad*mgva->rad - x[0]*x[0])/(meep::pi * mgva->rad*mgva->rad);
+  else if (mgva->dim == meep::D3)
+    w = meep::pi*(mgva->rad*mgva->rad - x[0]*x[0])/(4/3 * meep::pi * mgva->rad*mgva->rad*mgva->rad);
+  return ret * w;
+}
+#else
+static number matgrid_eps_func(int n, number *x, void *mgva_) {
+  matgrid_volavg *mgva = (matgrid_volavg *)mgva_;
+  // use a linear approximation for the material grid weights around the Yee grid point
+  double u_proj = tanh_projection(mgva->uval + mgva->ugrad_abs*x[0], mgva->beta, mgva->eta);
+  vector3 med1_eps_diag = mgva->med1_eps_diag;
+  vector3 med2_eps_diag = mgva->med2_eps_diag;
+  double eps1 = (med1_eps_diag.x + med1_eps_diag.y + med1_eps_diag.z)/3;
+  double eps2 = (med2_eps_diag.x + med2_eps_diag.y + med2_eps_diag.z)/3;
+  double eps_interp = (1-u_proj)*eps1 + u_proj*eps2;
+  double w = 0;
+  if (mgva->dim == meep::D1)
+    w = 1/(2*mgva->rad);
+  else if (mgva->dim == meep::D2)
+    w = 2*sqrt(mgva->rad*mgva->rad - x[0]*x[0])/(meep::pi * mgva->rad*mgva->rad);
+  else if (mgva->dim == meep::D3)
+    w = meep::pi*(mgva->rad*mgva->rad - x[0]*x[0])/(4/3 * meep::pi * mgva->rad*mgva->rad*mgva->rad);
+  return eps_interp * w;
+}
+static number matgrid_inveps_func(int n, number *x, void *mgva_) {
+  matgrid_volavg *mgva = (matgrid_volavg *)mgva_;
+  // use a linear approximation for the material grid weights around the Yee grid point
+  double u_proj = tanh_projection(mgva->uval + mgva->ugrad_abs*x[0], mgva->beta, mgva->eta);
+  vector3 med1_eps_diag = mgva->med1_eps_diag;
+  vector3 med2_eps_diag = mgva->med2_eps_diag;
+  double eps1 = (med1_eps_diag.x + med1_eps_diag.y + med1_eps_diag.z)/3;
+  double eps2 = (med2_eps_diag.x + med2_eps_diag.y + med2_eps_diag.z)/3;
```

Comment:
These could be stored directly in `mgva` since they are independent of `x`.

Reply:
Fixed.